### PR TITLE
Support for custom serializer

### LINKED ISSUE #178: Dependecy missing in 1.7.1
## Description of Bug

Kotlinxserialization is missing in 1.7.1 package which cause the emojiManager to fail loading and parsing the `emoji.json`
Here is the logs:
```
java.lang.NoClassDefFoundError: Failed resolution of: Lkotlinx/serialization/json/JsonKt;
	at io.wax911.emojify.initializer.EmojiInitializer.initEmojiData(EmojiInitializer.kt:44)
	at io.wax911.emojify.initializer.EmojiInitializer.initEmojiData$default(EmojiInitializer.kt:42)
	at io.wax911.emojify.initializer.EmojiInitializer.create(EmojiInitializer.kt:58)
	at io.wax911.emojify.initializer.EmojiInitializer.create(EmojiInitializer.kt:30)
	at androidx.startup.AppInitializer.doInitialize(AppInitializer.java:180)
	at androidx.startup.AppInitializer.discoverAndInitialize(AppInitializer.java:238)
	at androidx.startup.AppInitializer.discoverAndInitialize(AppInitializer.java:206)
	at androidx.startup.InitializationProvider.onCreate(InitializationProvider.java:45)
	at android.content.ContentProvider.attachInfo(ContentProvider.java:2404)
	at android.content.ContentProvider.attachInfo(ContentProvider.java:2374)
	at android.app.ActivityThread.installProvider(ActivityThread.java:7459)
	at android.app.ActivityThread.installContentProviders(ActivityThread.java:6976)
	at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6747)
	at android.app.ActivityThread.access$1500(ActivityThread.java:256)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2091)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loopOnce(Looper.java:201)
	at android.os.Looper.loop(Looper.java:288)
	at android.app.ActivityThread.main(ActivityThread.java:7870)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
Caused by: java.lang.ClassNotFoundException: Didn't find class "kotlinx.serialization.json.JsonKt" on path: DexPathList[[zip file "/data/app/~~NlkoSJsQPk2J_xmEXqaLdg==/io.wax911.emoji.sample-8O505Ni-SFZUXtbtx4Xlug==/base.apk"],nativeLibraryDirectories=[/data/app/~~NlkoSJsQPk2J_xmEXqaLdg==/io.wax911.emoji.sample-8O505Ni-SFZUXtbtx4Xlug==/lib/arm64, /system/lib64, /system_ext/lib64]]
	at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:259)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
	... 22 more
```

## Reproduction Steps

Implement the library via jitpack and start the app.

## To solve

In order to solve this issue there are multiple solution I have in mind:
- [`implementation(libs.jetbrains.kotlinx.serialization.json)`](https://github.com/AniTrend/android-emojify/blob/9d5d043f012efb8c601fd12233dd264044b41ae7/emojify/build.gradle.kts#L9) into `api(libs.jetbrains.kotlinx.serialization.json)`
- Allow user to specify its own parser/deparser to avoid having multiple lib doing the same thing in their project
- Not depend on parser lib

NB: Also any reason why minSdk is 23 ?

### PR DESCRIPTION
**Pull request by yoobi**
## Description

PR is related to issue: #178 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Enhancement (Improves existing functionality)

**IMPORTANT**: By submitting a patch, you agree to allow the project
owners to license your work under the terms of
the [Apache License 2.0](https://github.com/AniTrend/android-emojify/blob/develop/LICENSE.md).

### CODE REVIEW

**Comment by yoobi** [comment]
Small comment to say that since this is a breaking change, I've bump the version to 2.7.1, please tell me if it's alright with you

**Comment by yoobi** [comment]
There is also a possiblity of non-breaking change for user who use the EmojiInitializer. 

Inside the lib AndroidManifest:
```xml
<provider
    android:name="androidx.startup.InitializationProvider"
    android:authorities="${applicationId}.androidx-startup"
    android:exported="false"
    tools:node="merge">
    <meta-data
        android:name="io.wax911.emojify.initializer.EmojiInitializer"
        android:value="androidx.startup" />
</provider>
```

If user is using custom Deserializer
Inside the user AndroidManifest:
```xml
<provider
    android:name="androidx.startup.InitializationProvider"
    android:authorities="${applicationId}.androidx-startup"
    android:exported="false"
    tools:node="merge">
    <meta-data
        android:name="com.package.CustomEmojiInitializer"
        android:value="androidx.startup" />
    <meta-data
        android:name="io.wax911.emojify.initializer.EmojiInitializer"
        android:value="androidx.startup"
        tools:node="remove" />
</provider>
```

**Comment by wax911** [comment]
Please run `./gradlew :emojify:spotlessCheck` to check for any formatting issues and `./gradlew :emojify:spotlessApply` to apply auto fixable

**Comment by yoobi** [comment]
Sure ! It is done :)

I've noticed that you've tagged this changed as breaking but it is not

**Comment by wax911** [comment]
> Sure ! It is done :)
> 
> I've noticed that you've tagged this changed as breaking but it is not

Oh, I had seen this https://github.com/AniTrend/android-emojify/issues/178#issuecomment-1943478378 thus the breaking status

**Comment by yoobi** [comment]
Yes at first it was a breaking change, but I've refacto the code in order to change it as non-breaking. I've explained in this [comment](https://github.com/AniTrend/android-emojify/pull/179#issuecomment-1943553082)

File: gradle/version.properties
Comment (wax911) [inline gradle/version.properties:1]: We don't need to manually edit this file btw, a github action will draft a release based on the PR label and the `version.properties` file will be created along with the next PR

File: README.md [hunk 1]
```diff
@@ -177,6 +177,8 @@ dependencies {
 Don't know how to do that?? Take a look at
 the [application class example](./app/src/main/java/io/wax911/emojifysample/App.kt)
 
+**In order for EmojiInitilizer to work you need to add in your build.gradle `implementation "org.jetbrains.kotlinx:kotlinx-serialization-json:1.6.2"`
+
```
Comment (wax911) [inline README.md:181]: I liked the initial idea of having `api()` for gradle, any reason why you chose to advice consumers to add `implementation`? If someone uses a custom serialization they'd still be forced to add this into their project

File: gradle/version.properties
Comment (yoobi) [reply gradle/version.properties:1]: Oh ok ! Do you I need to revert it back to 1.7.1 ?

**Comment by wax911** [comment]
The PR looks solid, I just have one concern though, since the property annotation on the Model class are for `kotlinx.serialization` unless we make an interface/abstraction for the model and allow users to implement it and provide annotation for deserialization which we can substitute with the interface for usage. Otherwise some of the json property names don't automatically match the model property names. What are your thoughts?

File: README.md (hunk 1 above)
Comment (yoobi) [reply README.md:181]: Yes, the main difference between `api` and `implementation` is that when you use `api` it'll bring this library to the user project as well, but `implementation` is used the library will be only available in emojify.
There could be 2 kinds of "issue":
1. If user is using a different parsing library (like in my case) I'll have both Moshi and Kotlinx library in my project.
2. Users can have unexpected error if he uses a different version of kotlinx-serialization of the one used in emojify.

I have made a default KotlinxSerialization in order to not create breaking change, but even that is a bit of an issue as explained the in 2.

**Comment by yoobi** [comment]
I have made a couple test with MoshiDeserializer (I have made a CustomDeserializer in the app example so users could experiment with it) and even though there was no Moshi annotation on the `Emoji` model it did find the right data for `supports_fitzpatrick` and `supports_gender`. I could try with Gson or Jackson to make sure if we need to abstract the model or not

File: gradle/version.properties
Comment (wax911) [reply gradle/version.properties:1]: Yes please

**Comment by yoobi** [comment]
So I've added implementation for Jackson and Gson and they do work fine without the need to expose the Model.
I've made a change in the first emoji to try it out and they did return "true" for `supports_fitzpatrick` and `supports_gender`

**Comment by wax911** [comment]
Perfect, this is amazing! Could we just rename `AEmojiInitializer` to `AbstractEmojiInitializer`

**Review by wax911 — APPROVED** [review]

**Review by wax911 — APPROVED** [review]

**Comment by yoobi** [comment]
Hello ! I'm back with bad news :'(
I have tried to implement the new version in my project and it seems what we feared (https://github.com/AniTrend/android-emojify/pull/179#issuecomment-1951838672) is true and kotlinx-serialization is still needed

**Pull request merged**